### ISSUE TITLE
Respect assignment to namedtuple `__new__.__defaults__`

### ISSUE THREAD

**Issue by alexprengere** [astral-sh/ty#3268]
### Summary

I am trying to annotate a large Python codebase using ty as reference, I have many issues reported because of lack of handling of this `namedtuple` idiom:

```python
from collections import namedtuple

Version = namedtuple('Version', ['major', 'minor'])
Version.__new__.__defaults__ = (0, 0)

print(Version())
```
This reports:
```
error[missing-argument]: No arguments provided for required parameters `major`, `minor`
```
My question is simple: is there a change that this will be supported in ty at some point, or should I find another way to write this code, friendlier to type checkers?

### Version

0.0.29

**Comment by AlexWaygood** [comment]
If you want your namedtuples to be easily understood by type checkers, I would recommend using `typing.NamedTuple` rather than `collections.namedtuple`. Your `Version` class could be written like this, for example:

```py
from typing import NamedTuple

class Version(NamedTuple):
    major: int = 0
    minor: int = 0
```

and ty will understand it [just fine](https://play.ty.dev/d181e2ec-b7e8-4056-a0f1-10eb4ddc428c)

**Comment by alexprengere** [comment]
Thanks for the quick reply!

I need to re-check if the figures still hold, but I remember that `typing.NamedTuple` being slower than `collections.namedtuple`, being the reason to use the latter.

**Comment by AlexWaygood** [comment]
Hmm, slower in what way exactly? It's possible it might add a tiny amount of overhead when importing your module (the logic for _creating_ the namedtuple class itself is a bit more complex in `typing.NamedTuple` than it is in `collections.namedtuple`). But I believe that the class that's created at the end of the day is identical whether you're using `collections.namedtuple` or `typing.NamedTuple`, so I don't believe it would be any slower when it comes to actually _instantiating_ or _using_ the class. I might be wrong, though.

**Comment by AlexWaygood** [comment]
You can also get ty to understand your code by using the `defaults=` keyword to `namedtuple`, if you prefer it over `typing.NamedTuple`:

```py
from collections import namedtuple
Version = namedtuple("Version", "major minor", defaults=[0, 0])
Version()
Version(major=0, minor=0)
```

https://play.ty.dev/42cf48c4-8f7b-44c8-ac2e-5b232fdc15b1

**Comment by alexprengere** [comment]
Thanks!
* I double-checked on 3.14, I noticed 0 meaningful differences for instanciation/attribute access/attribute setting in terms of performance, but I think a few years ago this was not the case
* the `defaults` parameter was added back in 3.7, so the best course course of action is to stop relying on `__new__` for such a simple use case

So I will just use `default=` everywhere for now 😉 , thanks!

**Comment by carljm** [comment]
Given that there seem to be multiple preferable alternatives to this pattern, and that the way other type checkers support it is unsound (they just assume all fields have defaults if any tuple is assigned to `__defaults__`), I would be inclined to not support it in ty. But we can leave this issue open for now so it's findable for anyone who wants to present reasons why it should be supported.

**Comment by AlexWaygood** [comment]
> But we can leave this issue open for now so it's findable for anyone who wants to present reasons why it should be suported.

If there is very little chance that we will take any action here (and I think that's the case), I would prefer for us to just add an entry to our FAQs page on the docs, so that we can keep the issue tracker focussed on items where there is some chance that we might want to make a change in the future

**Comment by carljm** [comment]
I won't complain if you close it.

**Comment by AlexWaygood** [comment]
(I'll wait until we get a second report before adding it to FAQs -- currently it's really just an "AQ", given that we've only had one report so far 😄)
